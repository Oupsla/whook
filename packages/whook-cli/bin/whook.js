#! /usr/bin/env node

const path = require('path');
const run = require('../dist/index').default;
const PROJECT_SRC =
  process.env.PROJECT_SRC || path.join(process.cwd(), 'dist', 'index');
const { prepareEnvironment } = require(PROJECT_SRC);

run({ prepareEnvironment });
