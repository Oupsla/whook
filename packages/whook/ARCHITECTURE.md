[//]: # ( )
[//]: # (This file is automatically generated by the `jsarch`)
[//]: # (module. Do not change it elsewhere, changes would)
[//]: # (be overriden.)
[//]: # ( )
# Architecture Notes



## Server run

Whook exposes a `runServer` function to programmatically spawn
 its server. It is intended to be reusable and injectable so
 that projects can override the whole `whook` default behavior.

[See in context](./src/index.ts#L73-L77)



## Server preparation

Whook exposes a `prepareServer` function to create its server
 configuration. It takes eventually additional injections that
 would be required at a higher level and a
 [Knifecycle](https://github.com/nfroidure/knifecycle)
 containing the bootstrapped environment and allowing
 to complete and run the server.

[See in context](./src/index.ts#L143-L150)



### Root injections

* We need to inject `httpServer` and `process` to bring life to our
 *  server. We also inject `log` for logging purpose and custom other
 *  injected name that were required upfront.

[See in context](./src/index.ts#L164-L168)



## Server environment

The Whook `prepareEnvironment` function aims to provide the complete
 server environment without effectively planning its run. It allows
 to use that environment for CLI or build purposes. It also
 provides a chance to override some services/constants
 before actually preparing the server.

[See in context](./src/index.ts#L181-L187)



### `PWD` env var

The Whook server heavily rely on the process working directory
 to dynamically load contents. We are making it available to
 the DI system as a constant.

[See in context](./src/index.ts#L198-L202)



### `NODE_ENV` env var

Whook has different behaviors depending on the `NODE_ENV` value
 consider setting it to production before shipping.

[See in context](./src/index.ts#L206-L209)



### `WHOOK_PLUGINS` and `PROJECT_SRC`

Whook need to know where to look up for things like
 commands / handlers etc...

[See in context](./src/index.ts#L213-L216)



### Logging

Whook's default logger write to the NodeJS default console
 except for debugging messages where it uses the `debug`
 module so that you can set the `DEBUG` environment
 variable to `whook` and get debug messages in output.

[See in context](./src/index.ts#L219-L224)



### Initializers

Whook's embed a few default initializers proxied from
 `common-services`, `@whook/http-router` or its own
 `src/services` folder. It can be wrapped or overriden,
 at will, later in project's main file.

[See in context](./src/index.ts#L243-L248)



## Environment service

The `ENV` service add a layer of configuration over just using
 node's `process.env` value. Beware that `PWD` and `NODE_ENV` are
 guaranteed to be the exact same than the injected constants.
 It is up to you to decide upstream if you set them via the
 `process.env.NODE_ENV` and `process.cwd()` values or not.

[See in context](./src/services/ENV.ts#L8-L14)



## Base URL

The `BASE_URL` service is intended to provide a base URL where
 the API can be found at. It can be overriden directly via
 injecting it but it is useful to have a usable URL while
 debugging production environnement.

[See in context](./src/services/BASE_URL.ts#L7-L12)



### Environment isolation

Per default, Whook takes the process environment as is
 but since it could lead to leaks when building for
 AWS Lambda or Google Cloud Functions one can isolate
 the process env when building.

[See in context](./src/services/ENV.ts#L64-L69)



### `.env.NODE_ENV` files

You may need to keep some secrets out of your Git
 history. Whook uses `dotenv` to provide your such
 ability.

[See in context](./src/services/ENV.ts#L77-L81)



## `$autoload` service

The default Whook autoloader provides a simple way to
 load the constants, services and handlers of a Whook
 project automatically from the installed whook plugins.

[See in context](./src/services/_autoload.ts#L38-L42)



### Configuration auto loading

Loading the configuration files is done according to the `NODE_ENV`
 environment variable. It basically requires a configuration hash
 where the keys are Knifecycle constants.

Let's load the configuration files as a convenient way
 to create constants on the fly

[See in context](./src/services/_autoload.ts#L183-L190)



### Wrappers auto loading support

We cannot inject the `WRAPPERS` in the auto loader when
 it is dynamically loaded so giving a second chance here
 for `WRAPPERS` to be set.

[See in context](./src/services/_autoload.ts#L126-L130)



### API auto loading

We cannot inject the `API` in the auto loader since
 it is dynamically loaded so doing this during the auto
 loader initialization.

[See in context](./src/services/_autoload.ts#L110-L114)



### Constants

First of all the autoloader looks for constants in the
 previously loaded `CONFIGS` configurations hash.

[See in context](./src/services/_autoload.ts#L198-L201)



### Handlers map

Here, we build the handlers map needed by the router by injecting every
 handler required by the API.

[See in context](./src/services/_autoload.ts#L213-L216)



### Service/handler loading

Finally, we either require the handler/service module if
 none of the previous strategies applyed.

[See in context](./src/services/_autoload.ts#L243-L246)



#### WhookInitializerMap

Whook exports a `WhookInitializerMap` type to help you ensure yours are valid.

[See in context](./src/services/_autoload.ts#L33-L35)



### Service name mapping

In order to be able to substituate easily a service per another
 one can specify a mapping between a service and its substitution.

[See in context](./src/services/_autoload.ts#L170-L173)



#### WhookServiceMap

Whook exports a `WhookServiceMap` type to help you ensure yours are valid.

[See in context](./src/services/_autoload.ts#L29-L31)



### Initializer path mapping

In order to be able to load a service from a given path map
 one can directly specify a path to use for its resolution.

[See in context](./src/services/_autoload.ts#L276-L279)



## IP detection

If no `HOST` configuration is specified in dependencies nor in ENV,
 this service detects the machine host automagically.

[See in context](./src/services/HOST.ts#L7-L10)



## Port detection

If no `PORT` configuration is specified in dependencies nor in ENV,
 this service detects a free port automagically.

[See in context](./src/services/PORT.ts#L7-L10)



## Project dir

Whook needs to know the directory of the project under
 which he is running. It then uses this service to
 automatically detect it.

[See in context](./src/services/PROJECT_DIR.ts#L7-L12)



## Plugins paths

Whook auto loader can look for initializers in a list of
 plugins defined in the `WHOOK_PLUGINS` constant. This
 service computes the path where those plugins source are
 located allowing one to use services/handlers from it.

[See in context](./src/services/WHOOK_PLUGINS_PATHS.ts#L9-L15)

