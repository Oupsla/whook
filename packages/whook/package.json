{
  "name": "@whook/whook",
  "version": "4.0.0-alpha.11",
  "description": "Build strong and efficient REST web services.",
  "metapak": {
    "configs": [
      "main",
      "readme",
      "eslint",
      "babel",
      "jest",
      "jsarch",
      "jsdocs"
    ],
    "data": {
      "childPackage": true,
      "testsFiles": "'src/**/*.test.js'",
      "files": "'src/**/*.js'",
      "ignore": [
        "dist"
      ],
      "bundleFiles": [
        "dist/**/*.js"
      ]
    }
  },
  "main": "dist/index.js",
  "scripts": {
    "architecture": "jsarch 'src/**/*.js' > ARCHITECTURE.md && git add ARCHITECTURE.md",
    "changelog": "conventional-changelog -p angular -i CHANGELOG.md -s",
    "cli": "env NODE_ENV=${NODE_ENV:-cli}",
    "compile": "babel src --out-dir=dist",
    "cover": "npm run jest -- --coverage",
    "cz": "env NODE_ENV=${NODE_ENV:-cli} git cz",
    "doc": "echo \"# API\" > API.md; jsdoc2md 'src/**/*.js' >> API.md && git add API.md",
    "jest": "NODE_ENV=test jest",
    "lint": "eslint 'src/**/*.js'",
    "metapak": "metapak",
    "precz": "npm t && npm run lint && npm run metapak -- -s",
    "prettier": "prettier --write 'src/**/*.js'",
    "preversion": "npm t && npm run lint && npm run compile && npm run metapak -- -s",
    "start": "node bin/server.js",
    "test": "npm run jest"
  },
  "keywords": [
    "REST",
    "HTTP",
    "server",
    "web",
    "services",
    "OpenAPI"
  ],
  "author": {
    "name": "Nicolas Froidure",
    "email": "nicolas.froidure@insertafter.com",
    "url": "http://insertafter.com/en/index.html"
  },
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/nfroidure/whook.git"
  },
  "bugs": {
    "url": "https://github.com/nfroidure/whook/issues"
  },
  "homepage": "https://github.com/nfroidure/whook",
  "devDependencies": {
    "@babel/cli": "^7.2.3",
    "@babel/core": "^7.2.2",
    "@babel/plugin-proposal-object-rest-spread": "^7.3.1",
    "@babel/preset-env": "^7.3.1",
    "@babel/register": "^7.0.0",
    "babel-eslint": "^10.0.1",
    "babel-plugin-knifecycle": "^1.0.3",
    "eslint": "^5.13.0",
    "eslint-plugin-prettier": "^3.0.1",
    "jest": "^24.0.0",
    "jsarch": "^2.0.1",
    "jsdoc-to-markdown": "^4.0.1",
    "metapak": "^3.1.6",
    "metapak-nfroidure": "9.6.2",
    "prettier": "^1.16.3"
  },
  "dependencies": {
    "@whook/http-router": "^4.0.0-alpha.11",
    "@whook/http-server": "^4.0.0-alpha.11",
    "@whook/http-transaction": "^4.0.0-alpha.11",
    "common-services": "^5.0.3",
    "debug": "^4.1.1",
    "dotenv": "^8.0.0",
    "internal-ip": "^4.3.0",
    "knifecycle": "^6.0.0",
    "pkg-dir": "^4.2.0",
    "portfinder": "^1.0.21"
  },
  "contributors": [],
  "engines": {
    "node": ">=8.12.0"
  },
  "babel": {
    "presets": [
      [
        "@babel/env",
        {
          "targets": {
            "node": "8.12.0"
          }
        }
      ]
    ],
    "plugins": [
      "@babel/plugin-proposal-object-rest-spread",
      "babel-plugin-knifecycle"
    ]
  },
  "jest": {
    "coverageReporters": [
      "lcov",
      "html"
    ],
    "testPathIgnorePatterns": [
      "/node_modules/"
    ],
    "roots": [
      "<rootDir>/src"
    ]
  },
  "eslintConfig": {
    "extends": "eslint:recommended",
    "parserOptions": {
      "ecmaVersion": 2018,
      "sourceType": "module",
      "modules": true
    },
    "env": {
      "es6": true,
      "node": true,
      "jest": true,
      "mocha": true
    },
    "plugins": [
      "prettier"
    ],
    "rules": {
      "prettier/prettier": "error"
    }
  },
  "prettier": {
    "semi": true,
    "printWidth": 80,
    "singleQuote": true,
    "trailingComma": "all",
    "proseWrap": "always"
  },
  "files": [
    "dist/**/*.js",
    "LICENSE",
    "README.md",
    "CHANGELOG.md"
  ]
}
