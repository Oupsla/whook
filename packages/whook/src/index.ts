import Knifecycle, { constant, Services } from 'knifecycle';
import debug from 'debug';
import {
  initLogService,
  initTimeService,
  initRandomService,
  initDelayService,
  initProcessService,
  DEFAULT_LOG_ROUTING,
  LogService,
} from 'common-services';
import initHTTPRouter, {
  initErrorHandler,
  HTTPRouterProvider,
  HTTPRouterService,
} from '@whook/http-router';
import initHTTPTransaction, {
  WhookOperation,
  WhookRequest,
  WhookResponse,
  WhookHandler,
  HTTPTransactionService,
} from '@whook/http-transaction';
import initHTTPServer, {
  HTTPServerProvider,
  HTTPServerService,
} from '@whook/http-server';
import initPort from './services/PORT';
import initHost from './services/HOST';
import initEnv, { ENVService } from './services/ENV';
import initConfigs, { CONFIGSService, WhookConfig } from './services/CONFIGS';
import initProjectDir from './services/PROJECT_DIR';
import initWhookPluginsPaths, {
  WhookPluginsService,
  WhookPluginsPathsService,
} from './services/WHOOK_PLUGINS_PATHS';
import initAutoload, {
  HANDLER_REG_EXP,
  WhookWrapper,
  WhookServiceMap,
  WhookInitializerMap,
} from './services/_autoload';
import { noop, identity, compose, pipe } from './libs/utils';
import { OpenAPIV3 } from 'openapi-types';

export {
  noop,
  identity,
  compose,
  pipe,
  initAutoload,
  WhookServiceMap,
  WhookInitializerMap,
  ENVService,
  WhookPluginsService,
  WhookPluginsPathsService,
  CONFIGSService,
  WhookConfig,
  WhookOperation,
  WhookRequest,
  WhookResponse,
  WhookHandler,
  WhookWrapper,
  HTTPTransactionService,
  HTTPRouterProvider,
  HTTPRouterService,
  HTTPServerProvider,
  HTTPServerService,
};

export type WhookDefinition = {
  path: string;
  method: string;
  operation: OpenAPIV3.OperationObject;
};

/* Architecture Note #1: Server run
Whook exposes a `runServer` function to programmatically spawn
 its server. It is intended to be reusable and injectable so
 that projects can override the whole `whook` default behavior.
*/
export async function runServer<S = Services>(
  aPrepareEnvironment: typeof prepareEnvironment,
  aPrepareServer: typeof prepareServer,
  injectedNames: string[] = [],
): Promise<S> {
  try {
    const $ = await aPrepareEnvironment();
    const { ENV, log, ...services } = await aPrepareServer<{
      ENV: ENVService;
      log: LogService;
      services: any[];
    }>([...new Set([...injectedNames, 'ENV', 'log'])], $);
    if (ENV.DRY_RUN) {
      log('warning', 'ðŸŒµ - Dry run, shutting down now!');
      await $.destroy();
      return ({} as unknown) as S;
    }

    if (ENV.MERMAID_RUN) {
      const CONFIG_REG_EXP = /^([A-Z0-9_]+)$/;
      const MERMAID_GRAPH_CONFIG = {
        classes: {
          handlers: 'fill:#e7cdd2,stroke:#ebd4cb,stroke-width:1px;',
          config: 'fill:#d4cdcc,stroke:#ebd4cb,stroke-width:1px;',
          others: 'fill:#ebd4cb,stroke:#000,stroke-width:1px;',
        },
        styles: [
          {
            pattern: HANDLER_REG_EXP,
            className: 'handlers',
          },
          {
            pattern: CONFIG_REG_EXP,
            className: 'config',
          },
          {
            pattern: /^(.+)$/,
            className: 'others',
          },
        ],
        shapes: [
          {
            pattern: HANDLER_REG_EXP,
            template: '$0(($0))',
          },
          {
            pattern: CONFIG_REG_EXP,
            template: '$0{$0}',
          },
        ],
      };
      log('warning', 'ðŸŒµ - Mermaid graph generated, shutting down now!');
      process.stdout.write($.toMermaidGraph(MERMAID_GRAPH_CONFIG));
      await $.destroy();
      return ({} as unknown) as S;
    }

    return ({ ENV, log, $instance: $, ...services } as unknown) as S;
  } catch (err) {
    // eslint-disable-next-line
    console.error('ðŸ’€ - Cannot launch the process:', err.stack);
    //process.exit(1);
  }
}

/* Architecture Note #2: Server preparation
Whook exposes a `prepareServer` function to create its server
 configuration. It takes eventually additional injections that
 would be required at a higher level and a
 [Knifecycle](https://github.com/nfroidure/knifecycle)
 containing the bootstrapped environment and allowing
 to complete and run the server.
*/
/**
 * Runs the Whook server
 * @param {Array<String>} injectedNames
 * Root dependencies names to instanciate and return
 * @param {Knifecycle} $
 * The Knifecycle instance to use for the server run
 * @returns Object
 * A promise of the injected services
 */
export async function prepareServer<S = Services>(
  injectedNames: string[] = [],
  $: Knifecycle,
): Promise<S> {
  /* Architecture Note #2.1: Root injections
   * We need to inject `httpServer` and `process` to bring life to our
   *  server. We also inject `log` for logging purpose and custom other
   *  injected name that were required upfront.
   */
  const {
    log,
    ...services
  }: { log: LogService; services: any[] } = await $.run([
    ...new Set([...injectedNames, 'log', 'httpServer', 'process']),
  ]);

  log('warning', 'On air ðŸš€ðŸŒ•');

  return ({ log, ...services } as unknown) as S;
}

/* Architecture Note #3: Server environment
The Whook `prepareEnvironment` function aims to provide the complete
 server environment without effectively planning its run. It allows
 to use that environment for CLI or build purposes. It also
 provides a chance to override some services/constants
 before actually preparing the server.
 */
/**
 * Prepare the Whook server environment
 * @param {Knifecycle} $
 * The Knifecycle instance to set the various services
 * @returns Promise<Knifecycle>
 * A promise of the Knifecycle instance
 */
export async function prepareEnvironment(
  $: Knifecycle = new Knifecycle(),
): Promise<Knifecycle> {
  /* Architecture Note #3.1: `PWD` env var
  The Whook server heavily rely on the process working directory
   to dynamically load contents. We are making it available to
   the DI system as a constant.
   */
  const PWD = process.cwd();
  $.register(constant('PWD', PWD));

  /* Architecture Note #3.2: `NODE_ENV` env var
  Whook has different behaviors depending on the `NODE_ENV` value
   consider setting it to production before shipping.
   */
  const NODE_ENV = process.env.NODE_ENV || 'development';
  $.register(constant('NODE_ENV', NODE_ENV));

  /* Architecture Note #3.3: `WHOOK_PLUGINS` and `PROJECT_SRC`
  Whook need to know where to look up for things like
   commands / handlers etc...
   */
  $.register(constant('WHOOK_PLUGINS', ['@whook/whook']));

  /* Architecture Note #3.4: Logging
  Whook's default logger write to the NodeJS default console
   except for debugging messages where it uses the `debug`
   module so that you can set the `DEBUG` environment
   variable to `whook` and get debug messages in output.
   */
  $.register(constant('debug', debug('whook')));
  $.register(
    constant('logger', {
      // eslint-disable-next-line
      error: console.error.bind(console),
      // eslint-disable-next-line
      info: console.info.bind(console),
      // eslint-disable-next-line
      warning: console.error.bind(console),
    }),
  );
  $.register(constant('exit', process.exit));

  // Needed to avoid a dead lock
  // TODO: Remove when fixed that issue
  // https://github.com/nfroidure/knifecycle/issues/108
  $.register(constant('LOG_ROUTING', DEFAULT_LOG_ROUTING));

  /* Architecture Note #3.5: Initializers
  Whook's embed a few default initializers proxied from
   `common-services`, `@whook/http-router` or its own
   `src/services` folder. It can be wrapped or overriden,
   at will, later in project's main file.
   */
  [
    initLogService,
    initTimeService,
    initRandomService,
    initDelayService,
    initProcessService,
    initHTTPRouter,
    initHTTPTransaction,
    initHTTPServer,
    initErrorHandler,
    initPort,
    initHost,
    initEnv,
    initConfigs,
    initWhookPluginsPaths,
    initProjectDir,
  ].forEach($.register.bind($));

  return $;
}
