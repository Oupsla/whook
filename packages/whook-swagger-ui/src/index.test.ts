import { constant, initializer } from 'knifecycle';
import axios from 'axios';
import { prepareServer, prepareEnvironment } from '@whook/whook';
import initHTTPRouter from '@whook/http-router';
import wrapHTTPRouterWithSwaggerUI from '.';
import YError from 'yerror';

describe('wrapHTTPRouterWithSwaggerUI', () => {
  const HOST = 'localhost';
  const PORT = 8888;
  const BASE_PATH = '/v1';
  const API = {
    openapi: '3.0.2',
    info: {
      version: '1.0.0',
      title: 'Sample OpenAPI',
      description: 'A sample OpenAPI file for testing purpose.',
    },
    paths: {
      '/ping': {
        get: {
          operationId: 'getPing',
          summary: "Checks API's availability.",
          responses: {
            '200': {
              description: 'Pong',
              content: {
                'application/json': {
                  schema: {
                    type: 'object',
                    additionalProperties: false,
                    properties: {
                      pong: {
                        type: 'string',
                        enum: ['pong'],
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  };
  const logger = {
    info: jest.fn(),
    error: jest.fn(),
  };
  const debug = jest.fn();
  const time = jest.fn();
  const $autoload = jest.fn(async serviceName => {
    throw new YError('E_UNMATCHED_DEPENDENCY', serviceName);
  });
  let $;

  beforeEach(() => {
    logger.info.mockReset();
    logger.error.mockReset();
    debug.mockReset();
    time.mockReset();
    $autoload.mockClear();
  });

  beforeEach(async () => {
    $ = await prepareEnvironment();

    $.register(
      initializer(
        {
          name: '$autoload',
          type: 'service',
          options: { singleton: true },
        },
        async () => $autoload,
      ),
    );
    $.register(constant('BASE_PATH', BASE_PATH));
    $.register(constant('API', API));
    $.register(constant('NODE_ENV', 'test'));
    $.register(constant('DEV_ACCESS_TOKEN', 'oudelali'));
    $.register(constant('HOST', HOST));
    $.register(constant('WRAPPERS', []));
    $.register(
      constant('HANDLERS', {
        getPing: jest.fn(async () => ({
          status: 200,
          headers: { 'content-type': 'application/json' },
          body: { ping: 'pong' },
        })),
      }),
    );
    $.register(constant('logger', logger));
    $.register(constant('debug', debug));
    $.register(constant('time', time));
    $.register(constant('NODE_ENVS', ['test']));
  });

  it('should work', async () => {
    $.register(constant('PORT', PORT));
    $.register(wrapHTTPRouterWithSwaggerUI(initHTTPRouter));
    $.register(constant('CONFIG', {}));
    $.register(constant('NODE_ENV', 'test'));
    $.register(constant('DEBUG_NODE_ENVS', ['test']));
    $.register(
      constant('ENV', {
        DEV_MODE: '1',
      }),
    );

    time.mockReturnValue(new Date('2010-03-06T00:00:00Z').getTime());

    const { $instance } = await prepareServer(
      ['$instance', 'httpServer', 'process'],
      $,
    );
    const { status, headers, data } = await axios.get(
      `http://${HOST}:${PORT}${BASE_PATH}/ping`,
    );

    await $instance.destroy();

    expect({
      status,
      headers: {
        ...headers,
        // Erasing the Date header that may be added by Axios :/
        date: undefined,
        etag: undefined,
        'last-modified': undefined,
        server: undefined,
      },
      data,

      debugCalls: debug.mock.calls,
      logInfoCalls: logger.info.mock.calls,
      logErrorCalls: logger.error.mock.calls,
      autoloaderCalls: $autoload.mock.calls,
    }).toMatchSnapshot();
  });

  it('should serve Swagger HTML', async () => {
    $.register(constant('PORT', PORT + 2));
    $.register(wrapHTTPRouterWithSwaggerUI(initHTTPRouter));
    $.register(
      constant('CONFIG', {
        localURL: `http://${HOST}:${PORT + 2}`,
      }),
    );
    $.register(constant('NODE_ENV', 'test'));
    $.register(constant('DEBUG_NODE_ENVS', ['test']));
    $.register(
      constant('ENV', {
        DEV_MODE: '1',
      }),
    );

    time.mockReturnValue(new Date('2010-03-06T00:00:00Z').getTime());

    const { $instance } = await prepareServer(
      ['$instance', 'httpServer', 'process'],
      $,
    );
    const { status, headers, data } = await axios.get(
      `http://${HOST}:${PORT + 2}/docs`,
    );

    await $instance.destroy();

    expect({
      status,
      headers: {
        ...headers,
        // Erasing the Date header that may be added by Axios :/
        date: undefined,
        etag: undefined,
        'last-modified': undefined,
        server: undefined,
      },
      data,

      debugCalls: debug.mock.calls,
      logInfoCalls: logger.info.mock.calls,
      logErrorCalls: logger.error.mock.calls,
      autoloaderCalls: $autoload.mock.calls,
    }).toMatchSnapshot();
  });

  it('should be bypassed with no debug env', async () => {
    $.register(constant('PORT', PORT + 1));
    $.register(wrapHTTPRouterWithSwaggerUI(initHTTPRouter));
    $.register(constant('NODE_ENV', 'test'));
    $.register(constant('DEBUG_NODE_ENVS', []));
    $.register(constant('ENV', {}));

    time.mockReturnValue(new Date('2012-01-15T00:00:00Z').getTime());

    const { $instance } = await prepareServer(
      ['$instance', 'httpServer', 'process'],
      $,
    );

    await $instance.destroy();

    expect({
      debugCalls: debug.mock.calls,
      logInfoCalls: logger.info.mock.calls,
      logErrorCalls: logger.error.mock.calls,
      autoloaderCalls: $autoload.mock.calls,
    }).toMatchSnapshot();
  });
});
