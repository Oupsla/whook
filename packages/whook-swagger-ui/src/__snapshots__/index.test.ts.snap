// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapHTTPRouterWithSwaggerUI should be bypassed with no debug env 1`] = `
Object {
  "autoloaderCalls": Array [
    Array [
      "MAX_HEADERS_COUNT",
    ],
    Array [
      "KEEP_ALIVE_TIMEOUT",
    ],
    Array [
      "TIMEOUT",
    ],
    Array [
      "MAX_CONNECTIONS",
    ],
    Array [
      "PROCESS_NAME",
    ],
    Array [
      "SIGNALS",
    ],
    Array [
      "BUFFER_LIMIT",
    ],
    Array [
      "PARSERS",
    ],
    Array [
      "STRINGIFYERS",
    ],
    Array [
      "DECODERS",
    ],
    Array [
      "ENCODERS",
    ],
    Array [
      "QUERY_PARSER",
    ],
    Array [
      "TRANSACTIONS",
    ],
    Array [
      "uniqueId",
    ],
  ],
  "debugCalls": Array [
    Array [
      "üë£ - Logging service initialized.",
    ],
    Array [
      "‚åõ - Delay service initialized.",
    ],
    Array [
      "üìá - Process service initialized.",
    ],
    Array [
      "üí± - HTTP Transaction initialized.",
    ],
    Array [
      "üö¶ - HTTP Router initialized.",
    ],
    Array [
      "‚úÖ - Closing HTTP server.",
    ],
    Array [
      "‚úîÔ∏è - HTTP server closed!",
    ],
    Array [
      "‚è≥ - Cancelling pending timeouts:",
      0,
    ],
  ],
  "logErrorCalls": Array [
    Array [
      "üîÇ - Running in \\"test\\" environment.",
    ],
    Array [
      "üéôÔ∏è - HTTP Server listening at \\"http://localhost:8889\\".",
    ],
    Array [
      "On air üöÄüåï",
    ],
  ],
  "logInfoCalls": Array [],
}
`;

exports[`wrapHTTPRouterWithSwaggerUI should serve Swagger HTML 1`] = `
Object {
  "autoloaderCalls": Array [
    Array [
      "MAX_HEADERS_COUNT",
    ],
    Array [
      "KEEP_ALIVE_TIMEOUT",
    ],
    Array [
      "TIMEOUT",
    ],
    Array [
      "MAX_CONNECTIONS",
    ],
    Array [
      "PROCESS_NAME",
    ],
    Array [
      "SIGNALS",
    ],
    Array [
      "BUFFER_LIMIT",
    ],
    Array [
      "PARSERS",
    ],
    Array [
      "STRINGIFYERS",
    ],
    Array [
      "DECODERS",
    ],
    Array [
      "ENCODERS",
    ],
    Array [
      "QUERY_PARSER",
    ],
    Array [
      "TRANSACTIONS",
    ],
    Array [
      "uniqueId",
    ],
  ],
  "data": "<!-- HTML for static distribution bundle build -->
<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"UTF-8\\">
    <title>Swagger UI</title>
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"./swagger-ui.css\\" >
    <link rel=\\"icon\\" type=\\"image/png\\" href=\\"./favicon-32x32.png\\" sizes=\\"32x32\\" />
    <link rel=\\"icon\\" type=\\"image/png\\" href=\\"./favicon-16x16.png\\" sizes=\\"16x16\\" />
    <style>
      html
      {
        box-sizing: border-box;
        overflow: -moz-scrollbars-vertical;
        overflow-y: scroll;
      }

      *,
      *:before,
      *:after
      {
        box-sizing: inherit;
      }

      body
      {
        margin:0;
        background: #fafafa;
      }
    </style>
  </head>

  <body>
    <div id=\\"swagger-ui\\"></div>

    <script src=\\"./swagger-ui-bundle.js\\"> </script>
    <script src=\\"./swagger-ui-standalone-preset.js\\"> </script>
    <script>
    window.onload = function() {
      // Begin Swagger UI call region
      const ui = SwaggerUIBundle({
        url: \\"https://petstore.swagger.io/v2/swagger.json\\",
        dom_id: '#swagger-ui',
        deepLinking: true,
        presets: [
          SwaggerUIBundle.presets.apis,
          SwaggerUIStandalonePreset
        ],
        plugins: [
          SwaggerUIBundle.plugins.DownloadUrl
        ],
        layout: \\"StandaloneLayout\\"
      })
      // End Swagger UI call region

      window.ui = ui
    }
  </script>
  </body>
</html>
",
  "debugCalls": Array [
    Array [
      "üë£ - Logging service initialized.",
    ],
    Array [
      "‚åõ - Delay service initialized.",
    ],
    Array [
      "üìá - Process service initialized.",
    ],
    Array [
      "üí± - HTTP Transaction initialized.",
    ],
    Array [
      "üö¶ - HTTP Router initialized.",
    ],
    Array [
      "‚úÖ - Closing HTTP server.",
    ],
    Array [
      "‚úîÔ∏è - HTTP server closed!",
    ],
    Array [
      "‚è≥ - Cancelling pending timeouts:",
      0,
    ],
  ],
  "headers": Object {
    "cache-control": "max-age=3600",
    "connection": "close",
    "content-length": "1424",
    "content-type": "text/html",
    "date": undefined,
    "etag": undefined,
    "last-modified": undefined,
    "server": undefined,
  },
  "logErrorCalls": Array [
    Array [
      "üîÇ - Running in \\"test\\" environment.",
    ],
    Array [
      "üíÅ - Serving the public API docs: http://localhost:8890/docs?url=http%3A%2F%2Flocalhost%3A8890%2Fv1%2FopenAPI",
    ],
    Array [
      "üíÅ - Serving the private API docs: http://localhost:8890/docs?url=http%3A%2F%2Flocalhost%3A8890%2Fv1%2FopenAPI%3Faccess_token%3Doudelali",
    ],
    Array [
      "üéôÔ∏è - HTTP Server listening at \\"http://localhost:8890\\".",
    ],
    Array [
      "On air üöÄüåï",
    ],
  ],
  "logInfoCalls": Array [],
  "status": 200,
}
`;

exports[`wrapHTTPRouterWithSwaggerUI should work 1`] = `
Object {
  "autoloaderCalls": Array [
    Array [
      "MAX_HEADERS_COUNT",
    ],
    Array [
      "KEEP_ALIVE_TIMEOUT",
    ],
    Array [
      "TIMEOUT",
    ],
    Array [
      "MAX_CONNECTIONS",
    ],
    Array [
      "PROCESS_NAME",
    ],
    Array [
      "SIGNALS",
    ],
    Array [
      "BUFFER_LIMIT",
    ],
    Array [
      "PARSERS",
    ],
    Array [
      "STRINGIFYERS",
    ],
    Array [
      "DECODERS",
    ],
    Array [
      "ENCODERS",
    ],
    Array [
      "QUERY_PARSER",
    ],
    Array [
      "TRANSACTIONS",
    ],
    Array [
      "uniqueId",
    ],
  ],
  "data": Object {
    "ping": "pong",
  },
  "debugCalls": Array [
    Array [
      "üë£ - Logging service initialized.",
    ],
    Array [
      "‚åõ - Delay service initialized.",
    ],
    Array [
      "üìá - Process service initialized.",
    ],
    Array [
      "üí± - HTTP Transaction initialized.",
    ],
    Array [
      "üö¶ - HTTP Router initialized.",
    ],
    Array [
      "‚è≥ - Created a delay:",
      30000,
    ],
    Array [
      "‚è≥ - Cleared a delay",
    ],
    Array [
      "‚úÖ - Closing HTTP server.",
    ],
    Array [
      "‚úîÔ∏è - HTTP server closed!",
    ],
    Array [
      "‚è≥ - Cancelling pending timeouts:",
      0,
    ],
  ],
  "headers": Object {
    "connection": "close",
    "content-type": "application/json",
    "date": undefined,
    "etag": undefined,
    "last-modified": undefined,
    "server": undefined,
    "transaction-id": "0",
    "transfer-encoding": "chunked",
  },
  "logErrorCalls": Array [
    Array [
      "üîÇ - Running in \\"test\\" environment.",
    ],
    Array [
      "üíÅ - Serving the public API docs: http://localhost:8888/docs?url=http%3A%2F%2Flocalhost%3A8888%2Fv1%2FopenAPI",
    ],
    Array [
      "üíÅ - Serving the private API docs: http://localhost:8888/docs?url=http%3A%2F%2Flocalhost%3A8888%2Fv1%2FopenAPI%3Faccess_token%3Doudelali",
    ],
    Array [
      "üéôÔ∏è - HTTP Server listening at \\"http://localhost:8888\\".",
    ],
    Array [
      "On air üöÄüåï",
    ],
  ],
  "logInfoCalls": Array [
    Array [
      Object {
        "endInBytes": 135,
        "endOutBytes": 179,
        "endTime": 1267833600000,
        "errored": false,
        "id": "0",
        "ip": "127.0.0.1",
        "method": "GET",
        "protocol": "http",
        "reqHeaders": Object {
          "accept": "application/json, text/plain, */*",
          "connection": "close",
          "host": "localhost:8888",
          "user-agent": "axios/0.19.0",
        },
        "resHeaders": Object {
          "content-type": "application/json",
        },
        "startInBytes": 135,
        "startOutBytes": 0,
        "startTime": 1267833600000,
        "statusCode": 200,
        "url": "/v1/ping",
      },
    ],
  ],
  "status": 200,
}
`;
